src/lib/auth/auth0.ts

```diff
@@ -1,26 +1,483 @@
+
+    this.handleStorage = this.handleStorage.bind(this);
+  }
+
+  // Getters / Setters:
+
+  getAuthEventListenerHandler() {
+    return this.authEventListenerHandler;
+  }
+
+  setAppInfo(appInfo: AppInfo) {
+    this.appInfo = appInfo;
+  }
+
+  // Storage listeners:
+
+  initStorageSyncing() {
+    if (this.localStorageKey) {
+      window.addEventListener("storage", this.handleStorage);
+    } else {
+      console.warn(
+        "Calling `Othent.init` is a NOOP unless the `localStorageKey` option is used.",
+      );
+    }
+  }
+
+  stopStorageSyncing() {
+    window.removeEventListener("storage", this.handleStorage);
+  }
+
+  private handleStorage(event: StorageEvent) {
+    if (event.key !== this.localStorageKey) return;
+
+    if (event.newValue) {
+      this.restoreUserDetails();
+    } else {
+      this.logOut();
+    }
+  }
+
+  private persistUserDetails(userDetails: UserDetails | null) {
+    const { cookieKey, localStorageKey } = this;
+
+    if (cookieKey) {
+      if (userDetails) {
+        cookieStorage.setItem(cookieKey, JSON.stringify(userDetails));
+      } else if (cookieStorage.getItem(cookieKey) !== null) {
+        cookieStorage.removeItem(cookieKey);
+      }
+    }
+
+    if (localStorageKey) {
+      if (userDetails) {
+        const now = new Date();
+
+        const serializedUserDetails = JSON.stringify({
+          userDetails,
+          createdAt: now.toUTCString(),
+          expiredBy: new Date(
+            now.getTime() + this.refreshTokenExpirationMs,
+          ).toUTCString(),
+        } satisfies StoredUserDetails);
+
+        localStorage.setItem(localStorageKey, serializedUserDetails);
+      } else {
+        this.clearStoredUserDetails();
+      }
+    }
+  }
+
+  // `userDetails` setters:
+
+  private setUserDetails(userDetails: UserDetails | null, updateAuth = true) {
+    window.clearTimeout(this.userDetailsExpirationTimeoutID);
+
+    if (userDetails) {
+      this.userDetailsExpirationTimeoutID = window.setTimeout(
+        this.logOut,
+        this.refreshTokenExpirationMs,
+      );
+    }
+
+    const updatedAlreadyEmitted =
+      this.authEventListenerHandler.emit(userDetails);
+
+    if (!updatedAlreadyEmitted) {
+      // Only update this object (its ref) if something has actually changed, just in case some code in user land
+      // is actually relaying on this ref only changing if the data changes too:
+      this.userDetails = userDetails;
+    }
+
+    if (updateAuth) {
+      // We don't update `isAuthenticated`, `localStorage` or `cookieStorage` if `setUserDetails` was called from `restoreUserDetails`:
+      this.isAuthenticated = !!userDetails;
+      this.persistUserDetails(userDetails);
+    }
+
+    return userDetails;
+  }
+
+  private restoreUserDetails(userDetails?: UserDetails | null) {
+    let initialUserDetails = userDetails || null;
+
+    if (!initialUserDetails && this.localStorageKey) {
+      try {
+        const storedUserDetails = JSON.parse(
```

Comment:
This could be a security risk if an attacker gets access to the stored user data (personal information) or manipulates the stored values (depending how the app works with them). It might be worth mentioning these values should be validated and/or sanitized.

Reply:
Added:

> The value you get when calling `getUserDetails()` or when subscribing to the `auth` event before authenticating by calling `connect()` or `requireAuth()` (either manually or automatically by setting `autoConnect = "eager"`), will be (in this order of priority):
> 
> - The value you pass as `initialUserDetails`, if any. 
> - The value stored in `localStorage`, if you set `persistLocalStorage`.
> - `null` otherwise.
> 
> `@othent/kms` does not performs any type of validation when reading the user details from these sources, and therefore you should treat that data as untrusted and make sure you sanitize it appropriately.
> 
> Failing to do that could result in your application being vulnerable to XSS if an attacker manages to tamper with these values.